src/state/workload.rs

```diff
@@ -632,14 +633,83 @@ pub struct WorkloadStore {
     // is simpler (and only requires a channelsize of 1)
     insert_notifier: Sender<()>,
 
-    /// byAddress maps workload network addresses to workloads
-    by_addr: HashMap<NetworkAddress, Arc<Workload>>,
-    /// byUid maps workload UIDs to workloads
+    /// by_addr maps workload network addresses to workloads
+    by_addr: HashMap<NetworkAddress, WorkloadByAddr>,
+    /// by_uid maps workload UIDs to workloads
     pub(super) by_uid: HashMap<Strng, Arc<Workload>>,
     // Identity->Set of UIDs. Only stores local nodes
     node_local_by_identity: HashMap<WorkloadIdentity, HashSet<Strng>>,
 }
 
+#[derive(Debug)]
+/// WorkloadByAddr is a small wrapper around a single or multiple Workloads
+/// We split these as in the vast majority of cases there is only a single one, so we save vec allocation.
+enum WorkloadByAddr {
+    Single(Arc<Workload>),
+    Many(Vec<Arc<Workload>>),
+}
+
+impl WorkloadByAddr {
+    // insert adds the workload
+    pub fn insert(&mut self, w: Arc<Workload>) {
+        match self {
+            WorkloadByAddr::Single(workload) => {
+                *self = WorkloadByAddr::Many(vec![workload.clone(), w]);
+            }
+            WorkloadByAddr::Many(v) => {
+                v.push(w);
+            }
+        }
+    }
+    // remove_uid mutates the address to remove the workload referenced by the UID.
+    // If 'true' is returned, there is no workload remaining at all
+    pub fn remove_uid(&mut self, uid: Strng) -> bool {
+        match self {
+            WorkloadByAddr::Single(wl) => {
+                // Remove it if the UID matches, else do nothing
+                wl.uid == uid
+            }
+            WorkloadByAddr::Many(ws) => {
+                ws.retain(|w| w.uid != uid);
+                match ws.as_slice() {
+                    [] => true,
+                    [wl] => {
+                        // We now have one workload, transition to Single
+                        *self = WorkloadByAddr::Single(wl.clone());
```

Comment:
Is this even worth it? It's more correct, but wonder if a high churn scenario could make this expensive. Maybe it's faster to just let a `::Many` stay `::Many`?

Reply:
Oh but on the other hand, large cluster with high churn could mean frequently overlapping ips...

Reply:
Either case is probably expensive, the vec is probably going to be resized if we insert, so making a new many vs inserting 1 is probably likely to be expensive either way.

I wouldn't have done the optimization at all if it wasn't for the fact that you can only get into overlapping IP case with an invalid configuration. So the `many` is the "slow" path essential

Reply:
Ack - makes sense